caas/kubernetes/provider/k8swatcher.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v1"
+	"k8s.io/client-go/pkg/watch"
+
+	"github.com/juju/juju/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+// kubernetesWatcher reports changes to kuernetes
+// resources. A native kubernetes watcher is passed
+// in to generate change events from the kubernetes
+// model. These events are consolidated into a Juju
+// notification watcher event.
+type kubernetesWatcher struct {
+	catacomb catacomb.Catacomb
+
+	out       chan struct{}
+	name      string
+	k8watcher watch.Interface
+}
+
+func newKubernetesWatcher(wi watch.Interface, name string) (*kubernetesWatcher, error) {
+	w := &kubernetesWatcher{
+		out:       make(chan struct{}),
+		k8watcher: wi,
+		name:      name,
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, err
+}
+
+const sendDelay = 1 * time.Second
+
+func (w *kubernetesWatcher) loop() error {
+	defer close(w.out)
+
+	next := time.After(sendDelay)
+	eventReceived := false
+	// Set out now so that initial event is sent immediately.
+	out := w.out
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			w.k8watcher.Stop()
+			return tomb.ErrDying
+		case <-next:
+			next = time.After(sendDelay)
```

Comment:
rather than waking up every second to probably do nothing, I'd suggest you only start the timer when a w.k8watcher result is received, and there's no timer already active. so something like:

```go
var delayCh <-chan time.Time
for {
    select {
    ...
    case evt := <-w.k8watcher.ResultChan():
        ...
        if delayCh == nil {
            delayCh = time.After(sendDelay)
        }
    case <-delayCh:
        out = w.out
    case out <- struct{}{}:
        out = nil
        delayCh = nil
    ...
    }
}
```

Reply:
good idea